Holidayselector & Holiday systems in shared

IProduceWidgets opened the pull request:
<!-- Guidelines: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

## About the PR
<!-- What did you change? -->
I just wanted to add a selector man.

The HolidaySelector works on entity tables and is an AllSelector which only works on holidays.

Moved all the holiday stuff to shared so I could access it, with the exception of a couple things that are now in the ServerHolidaySystem instead

## Why / Balance
<!-- Discuss how this would affect game balance or explain why it was changed. Link any relevant discussions or issues. -->
I sure hope you don't use holidays for balance >.>

## Technical details
<!-- Summary of code changes for easier review. -->
Most everything was pretty much cut and paste into shared.
I changed a bunch of the namespace to reflect that,

I kept the bulk of the christmas stuff in server because it seemed annoying to move since it messes with other server systems, and also some of the components list secrets.

I left the stuff that interacts with gameticking and chat messaging in server because I think I have to.

## Media
<!-- Attach media if the PR makes ingame changes (clothing, items, features, etc). 
Small fixes/refactors are exempt. Media may be used in SS14 progress reports with credit. -->

## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [x] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [x] I have added media to this PR or it does not require an ingame showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

## Breaking changes
<!-- List any breaking changes, including namespaces, public class/method/field changes, prototype renames; and provide instructions for fixing them.
This will be posted in #codebase-changes. -->
You may need to change your references for the Holiday System to Shared instead of Server.

**Changelog**
<!-- Add a Changelog entry to make players aware of new features or changes that could affect gameplay.
Make sure to read the guidelines and take this Changelog template out of the comment block in order for it to show up.
Changelog must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog. -->
<!--
:cl:
- add: Added fun!
- remove: Removed fun!
- tweak: Changed fun!
- fix: Fixed fun!
-->
:cl:
- add: a Holiday conditioned entity table selector.

lzk228 commented:
i was thinking of adding holiday things to mail a few days ago and here it is possible now🥹

slarticodefast commented on Content.Shared/Holiday/Christmas/LimitedItemGiverComponent.cs:
```diff
@@ -1,13 +1,13 @@
 /// <summary>
 /// This is used for granting items to lucky souls, exactly once.
 /// </summary>
-[RegisterComponent, Access(typeof(LimitedItemGiverSystem))]
```
Why remove that?
Also moving all the components to shared won't do much alone without also networking them and moving the corresponding system to shared. I would leave them as they are for now in this PR and move them one by one.

IProduceWidgets replied:
Eh, sure. I was hopeful I could move the system out of shared too, but I think that's gonna be a bigger job, and there's maybe some reasons why parts of it shouldn't be to keep secrets from clients.

slarticodefast commented on Content.Shared/Holiday/HolidaySystem.cs:
```diff
@@ -1,30 +1,29 @@
-namespace Content.Server.Holiday
+namespace Content.Shared.Holiday
 {
     public sealed class HolidaySystem : EntitySystem
```
```suggestion
    public abstract partial class SharedHolidaySystem : EntitySystem
```
naming convention
rename the file accordingly
the server and client systems inheriting should be simply named `HolidaySystem`

slarticodefast commented on Content.Shared/Holiday/HolidaySystem.cs:
```diff
@@ -1,30 +1,29 @@
+        [Dependency] private readonly INetManager _net = default!;
         [Dependency] private readonly SharedAppearanceSystem _appearance = default!;
 
         [ViewVariables]
```
not your fault, but what the hell is this doing here?

slarticodefast commented on Content.Shared/Holiday/HolidaySystem.cs:
```diff
@@ -51,19 +50,14 @@ public void RefreshCurrentHolidays()
             foreach (var holiday in _currentHolidays)
             {
-                holiday.Celebrate();
+                if (_net.IsServer)
```
Move outside the loop.
Or even better, make this a virtual function and override it on the server. Same for `DoGreet`. That way they can be called from shared.

slarticodefast commented on Content.Shared/Holiday/HolidaySystem.cs:
```diff
@@ -1,30 +1,29 @@
+        public List<HolidayPrototype> CurrentHolidays => _currentHolidays;
 
         [ViewVariables]
         private bool _enabled = true;
+        public bool Enabled => _enabled;
```
wouldn't it be simpler to make it protected?

slarticodefast reviewed (changes requested):
You will need to network `DateTime.Now` from the server to the client or they may disagree depending on your computer's time settings.

EmoGarbage404 reviewed (changes requested):
instead of moving all of the holiday code into shared, just send a network message to clients containing the list of CurrentHolidays. Then all you need is to put the API in a shared system.

IProduceWidgets commented:
> You will need to network `DateTime.Now` from the server to the client or they may disagree depending on your computer's time settings.

I did so with a couple pairs of events now for both DateTime.Now and the ccvar for HolidayEnabled.

Not sure I used the right approach though.

IProduceWidgets commented:
> instead of moving all of the holiday code into shared, just send a network message to clients containing the list of CurrentHolidays. Then all you need is to put the API in a shared system.

That probably would have been easier yeah. 
Alas, my mistress Caffeine took me away. We here now.
Bright side, you can check against the ChineseLunisolarCalendar now. uh,

EmoGarbage404 commented on Content.Shared/EntityTable/EntitySelectors/HolidaySelector.cs:
```diff
@@ -0,0 +1,38 @@
+/// <summary>
+/// If the given Holiday is active, Gets spawns from all of the child selectors
+/// </summary>
+public sealed partial class HolidaySelector : EntityTableSelector
```
this really would ideally be handled by the condition system i have outlined in https://github.com/space-wizards/space-station-14/pull/36819.

not that this is particularly bad, i just think its a bit of a clunky way to do it with the lack of proper condition support.

IProduceWidgets replied:
Sounds like a plan. Yoinked the selector, but Ill have to wait on this and that PR to merge to make it due to shared.

IProduceWidgets commented:
Hold merge for a sec, I am dummy with network events and race conditions.

IProduceWidgets commented:
Kay, think I fixed that.

I do get these on connect, but Idk maybe its fine?
![image](https://github.com/user-attachments/assets/48212a76-8db7-4b45-8e02-de9f4edf84a8)

Let me know if Im just using network events wrong, I fixed the issue by just ending the method that raised the event I need the server to respond to to continue, and then having it call the second half of what it once was. This seems to have fixed the race issue, but It does not make me feel warm and cuddly.
More details in the form of a question:
https://discord.com/channels/310555209753690112/560845886263918612/1365797321580871692

EmoGarbage404 commented on Content.Shared/Holiday/SharedHolidaySystem.cs:
```diff
@@ -0,0 +1,120 @@
+                return;
+            }
+
+            RaiseNetworkEvent(new RequestWhatDateItIsEvent());
+
+            // I think I have to break the method here to let the network event resolve.
+            var now = _currentDate;
+            RefreshCurrentHolidaysPart2(now);
+        }
+
+        public void RefreshCurrentHolidaysPart2(DateTime then)
+        {
+            var now = _currentDate;
+
+            // The following is intended to interrogate race conditions since the first part of this method needs to let the server get the date and set it for shared.
+            if (_net.IsServer)
+                Log.Warning($"the value for now on Server is {now} and was {then}");
```
this doesn't work. the second function is just going to compute immediately before the network events hit anything.

in general, why are you bothering raising these events on the server? The server should just precalculate all of this and then the client, on connect, should send a message to the server to receive the data. i don't understand what all of this is for.

IProduceWidgets replied:
How do you await the response of a network event? It didn't work until I did this, but being honest I'm not sure why. 

But to answer your question; because for some reason we can hotload holidays.

IProduceWidgets replied:
I dunno, Ill sleep on it and try again.
